_posts/2025-07-14-OpenSearch-Approximation-Framework.md

```diff
@@ -0,0 +1,225 @@
+
+![big5 desc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/big5-desc_sort_timestamp-query-with-approximation.png)
+
+### http_logs: desc_sort_timestamp query
+
+Seen improvement from **~280ms** to **~15ms**
+
+![http_logs desc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-desc_sort_timestamp-query-without-approximation.png)
+
+![http_logs desc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-desc_sort_timestamp-query-with-approximation.png)
+
+### http_logs: asc_sort_timestamp query
+
+Seen improvement from **~15ms** to **~8ms**
+
+![http_logs asc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-asc_sort_timestamp-query-without-approximation.png)
+
+![http_logs asc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-asc_sort_timestamp-query-with-approximation.png)
+
+
+Further improvements, particularly in sort query performance, were observed in the `3.1` release and are shared in detail on the [3.1.0 release issue](https://github.com/opensearch-project/opensearch-build/issues/5487#issuecomment-2989202040).
+ 
+This result comes from a single-segment test for `desc_sort_timestamp`. From this [comment](https://github.com/opensearch-project/OpenSearch/pull/18439#issuecomment-2961325856) with a force merge to a single segment, using a optimized custom BKD walk (intersectRight), P90 latency dropped dramatically from **450624** ms to **7.86** ms. This improvement is made possible by the Approximation Framework, which enables early termination during BKD traversal, efficiently collecting the most relevant documents with minimal overhead.
+
+As shown in this [comment](https://github.com/opensearch-project/OpenSearch/pull/18439#issuecomment-2942212895), which captures benchmark results from the development phase, several query types showed significant improvements: `http_logs desc_sort_size` saw over **80%** improvement, `http_logs`   `desc_sort_timestamp` improved by more than **80%**, and `asc_sort_timestamp` achieved over **80.55%** improvement in performance.
+
+## Upcoming Improvements:
+
+Here is high level [META issue](https://github.com/opensearch-project/OpenSearch/issues/18619) for next set of enhancements related to Approximation Framework.
+
+The following are some of the upcoming improvements planned for the 3.2.0 release, including extending the Approximation Framework to support all numeric types (Related [PR](https://github.com/opensearch-project/OpenSearch/pull/18530)). Additional performance gains have also been observed in range and sort queries, particularly when tested on the skewed http_logs dataset.
+
+![http_logs performance comparison](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-performance-comparison.png)
```

Comment:
This organization is a bit unclear: the second sentence on line 185 seems to be a performance result. And the next sections (lines 189-213) seem to also be performance results. Should they be moved to before "Upcoming improvements"?

Reply:
Sure I will take a look. Also FYI we are an RFC for an upcoming enhancement related to this framework https://github.com/opensearch-project/OpenSearch/issues/18784. I would like to push a commit mentioning this. Thanks